src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2109 @@
+	} else {
+		draw_layer_tiles(bitmap, cliprect, tilemap, scroll_idx, is_overlay_rendered);
+	}
+}
+
+void sega_9h0_0008_state::draw_layer_tiles(bitmap_rgb32 &bitmap, const rectangle &cliprect, const uint16_t *tilemap, const uint8_t scroll_idx, const bool is_overlay_rendered)
+{
+	// Bitplane area =  1024x512 (64x32 tiles), matches tile data mapping entries (64 * 32 = 0x800).
+	//  Visible area =   704x480 (44x30 tiles), offset by +10 tiles in x-axis and +2 tiles in y-axis.
+	//  Tilemap area = 1024x1024 (64x64 tiles), all these off-screen tiles must be considered for scrolling.
+	for (size_t offset = 0; offset < 0x2000/2; offset++) {
+		int32_t y = ((offset / 64) - 2) * 16;
+		int32_t x = ((offset % 64) - 10) * 16;
+
+		uint16_t scroll_x_idx = scroll_idx;
+		uint16_t scroll_y_idx = scroll_idx;
+		uint8_t entry_size = 4; // 2 int16_t values, 1 for each tile layer
+		if (y >= 0 && y < (0x90 / entry_size) * 16) {
+			if ((m_video_regs[0x10/4] & 0x20) == 0x20) {
+				scroll_x_idx += y/16 * 2;
+			}
+		}
+		if (x >= 0 && x < (0xb4 / entry_size) * 16) {
+			if ((m_video_regs[0x10/4] & 0x40) == 0x40) {
+				scroll_y_idx += x/16 * 2;
+			}
+		}
+
+		x = scroll_x(x, scroll_x_idx);
+		y = scroll_y(y, scroll_y_idx);
+
+		uint16_t tile = tilemap[BYTE_XOR_BE(offset)];
+		uint16_t tile_transform = (tile & 0xf000) >> 12;
+		uint16_t flip_x = (tile_transform & 1) != 0;
+		uint16_t flip_y = (tile_transform & 2) != 0;
+		bool is_overlay = (tile_transform & 4) != 0;
+		if (is_overlay != is_overlay_rendered) {
+			continue;
+		}
+
+		uint16_t tile_idx = tile & 0x7ff;
+		bool is_sprite_tile_data_used = tile & 0x800;
+		gfx_element *gfx = m_gfxdecode->gfx(is_sprite_tile_data_used ? 1 : 0);
+		const pen_t *paldata = m_palette->pens() + gfx->colorbase() + gfx->granularity() * (7 % gfx->colors());
+		const uint8_t scale = m_scale;
+		drawgfxzoom_with_pixel_op(gfx, bitmap, cliprect,
+			tile_idx,
+			flip_x,
+			flip_y,
+			x * m_scale,
+			y * m_scale,
+			0x10000 * m_scale,
+			0x10000 * m_scale,
+			[scale, paldata](int32_t y, const std::unique_ptr<rgb_t[]> &palcache, uint32_t &destp, const uint8_t &srcp) {
+				if (srcp != 0) {
+					rgb_t color = palcache[y / scale * 0x100 + srcp];
+					if (color == 0) {
+						color = paldata[srcp];
+					}
+					destp = color;
+				}
+			});
+	}
+}
+
+/**
+ * Similar to gfx_element::drawgfxzoom_core(), but uses a pixel_op that depends on the current y-position to render,
+ * in order to apply dedicated palettes for each scanline. These palettes are cached and passed directly
+ * as a parameter instead of being captured in the pixel_op lambda, to avoid degraded performance from large copies.
+ */
+template <typename BitmapType, typename FunctionClass>
+inline void sega_9h0_0008_state::drawgfxzoom_with_pixel_op(gfx_element *gfx, BitmapType &dest, const rectangle &cliprect, uint32_t code, int flipx, int flipy, int32_t destx, int32_t desty, uint32_t scalex, uint32_t scaley, FunctionClass pixel_op)
```

Comment:
Probably better to make `pixel_op` a universal reference to make it easier for the compiler to avoid unnecessary copies, like `FunctionClass &&pixel_op`.